core/src/main/java/org/apache/iceberg/BaseFilesTable.java

```diff
@@ -140,42 +143,92 @@ protected CloseableIterable<FileScanTask> doPlanFiles() {
+            DataFile.UPPER_BOUNDS);
+
     private final FileIO io;
     private final Map<Integer, PartitionSpec> specsById;
     private final ManifestFile manifest;
-    private final Schema schema;
+    private final Schema dataTableSchema;
+    private final Schema projection;
 
     ManifestReadTask(
         Table table,
         ManifestFile manifest,
-        Schema schema,
+        Schema projection,
         String schemaString,
         String specString,
         ResidualEvaluator residuals) {
       super(DataFiles.fromManifest(manifest), null, schemaString, specString, residuals);
       this.io = table.io();
       this.specsById = Maps.newHashMap(table.specs());
       this.manifest = manifest;
-      this.schema = schema;
+      this.dataTableSchema = table.schema();
+      this.projection = projection;
     }
 
     @Override
     public CloseableIterable<StructLike> rows() {
-      return CloseableIterable.transform(manifestEntries(), file -> (StructLike) file);
+      if (projection.findColumnName(DataFile.READABLE_METRICS.fieldId()) == null) {
+        return CloseableIterable.transform(files(projection), file -> (StructLike) file);
+      } else {
+        Schema fileProjection = TypeUtil.selectNot(projection, READABLE_METRICS_FIELD_IDS);
+        Schema minProjection =
+            TypeUtil.joinCommon(fileProjection, MIN_PROJECTION_FOR_READABLE_METRICS);
+        return CloseableIterable.transform(files(minProjection), this::withReadableMetrics);
+      }
     }
 
-    private CloseableIterable<? extends ContentFile<?>> manifestEntries() {
+    private CloseableIterable<? extends ContentFile<?>> files(Schema fileProjection) {
       switch (manifest.content()) {
         case DATA:
-          return ManifestFiles.read(manifest, io, specsById).project(schema);
+          return ManifestFiles.read(manifest, io, specsById).project(fileProjection);
         case DELETES:
-          return ManifestFiles.readDeleteManifest(manifest, io, specsById).project(schema);
+          return ManifestFiles.readDeleteManifest(manifest, io, specsById).project(fileProjection);
         default:
           throw new IllegalArgumentException(
               "Unsupported manifest content type:" + manifest.content());
       }
     }
 
+    private StructLike withReadableMetrics(ContentFile<?> file) {
+      int structSize = projection.columns().size();
+      Map<String, StructLike> metrics =
+          MetricsUtil.readableMetricsMap(
+              dataTableSchema, file, readableMetricsProjection(projection));
+      return new ContentFileStructWithMetrics(structSize, (StructLike) file, metrics);
+    }
+
+    // Handles projections for readable metrics struct
+    private Map<Integer, Integer> readableMetricsProjection(Schema projectedSchema) {
+      Map<Integer, Integer> projectionMap = Maps.newHashMap();
+      Type type = projectedSchema.findType(DataFile.READABLE_METRICS_VALUE.valueId());
+      if (type != null) {
```

Comment:
I don't think `type` can be null. We don't have to add more validation as it is checked above. If the type is null, the projection map will be empty, which will break `ReadableMetricsStruct` later. Seems like we can simply remove the if statement.

Reply:
You are right, removed.